evm/evm-codec/src/codecs/array.ts

```diff
@@ -0,0 +1,85 @@
+import { Codec, WORD_SIZE } from "../codec";
+import { Sink } from "../sink";
+import { Src } from "../src";
+
+export class ArrayCodec<const T> implements Codec<readonly T[]> {
+  public readonly isDynamic = true;
+
+  constructor(public readonly item: Codec<T>) {}
+
+  encode(sink: Sink, val: T[]) {
+    sink.dynamicOffset(val.length);
+    for (let i = 0; i < val.length; i++) {
+      this.item.encode(sink, val[i]);
+    }
+    sink.increaseSize(WORD_SIZE);
+    sink.endDynamic();
+  }
+
+  decode(src: Src): T[] {
+    const offset = src.u32();
+
+    src.safeJump(offset);
+    const len = src.u32();
+
+    const tmpSrc = src.slice(offset + WORD_SIZE);
+    const val = new Array(len);
+    for (let i = 0; i < val.length; i++) {
+      val[i] = this.item.decode(tmpSrc);
+    }
+    src.jumpBack();
+    return val;
+  }
+}
+
+export class FixedArrayCodec<const T> implements Codec<readonly T[]> {
+  public isDynamic: boolean;
```

Comment:
How "fixed" array can be "dynamic"? What does it mean? Why not to make separate struct?

Reply:
https://docs.soliditylang.org/en/latest/abi-spec.html#formal-specification-of-the-encoding

A type string[2], for example would be considered dynamic. Same as tuple (int,int,string). From the user's POV it's still a fixed array size, only the parsing logic is different, so imo it should stay as a same struct

Reply:
I'll rename it to FixedSizeArray to avoid confusion